### PR TITLE
FIX: refactor so biasedurn works with numpy<1.17

### PR DESCRIPTION
**Pull request by mckib2**
#### What does this implement/fix?

- removes headers with no implementations (it was a good try, but it didn't work)
- rely on function pointers to yield random numbers instead of bit generators
    - additional benefit of making `stocR.h` even smaller
- conditionally build `biasedurn.pyx` based on numpy version

#### Additional information

I've tested with numpy==1.16.5, will rebuild with latest numpy to make sure that works too

_EDIT: don't merge yet, working through newer numpy issues now_

### CODE REVIEW

**Comment by mckib2** [comment]
@mdhaber This is looking good to me: it builds and passes tests for numpy==1.16.5 and 1.19 (on my machine at least...).  Please look over and let me know if you have any concerns

**Comment by mdhaber** [comment]
Let's try it! Thank you!

**Comment by mdhaber** [comment]
This didn't build for me, even after adding what you suggested in comments, so I reverted in case someone wants to review the PR while we get this sorted out. You should be able to get back to where you left off easily - just create a backup branch of what you have locally, go back to the remote branch and pull the latest, then revert the revert `git revert 7f7c4782de7c645f2ec3e640f0ceb91b5cb57ed4`. Thank you for your help!

**Comment by mckib2** [comment]
Okay, sorry for all the build errors.  I'll have to do more testing tonight.  Were you trying to build on Windows?  Like I mentioned on the main PR, I think we were close with that last one, I'll be interested to see what build errors you were running into

(this is where the "it works on my machine" becomes frustrating...)

**Comment by mdhaber** [comment]
Actually, build is fine; there was just an error running the test.  I think it's the same as in CI:
```
(scipydev) C:\Users\mhaberla\Desktop\scipy>python runtests.py -v -t scipy/stats/tests/test_discrete_basic.py
Building, see build.log...
Build OK (0:00:32.034864 elapsed)
================================================= test session starts =================================================
platform win32 -- Python 3.7.9, pytest-6.1.1, py-1.9.0, pluggy-0.13.1 -- C:\Users\mhaberla\Anaconda3\envs\scipydev\python.exe
cachedir: .pytest_cache
rootdir: C:\Users\mhaberla\Desktop\scipy, configfile: pytest.ini
collected 282 items

scipy\stats\tests\test_discrete_basic.py::test_discrete_basic[bernoulli-arg0-True] PASSED                        [  0%]
scipy\stats\tests\test_discrete_basic.py::test_discrete_basic[betabinom-arg1-True] PASSED                        [  0%]
scipy\stats\tests\test_discrete_basic.py::test_discrete_basic[binom-arg2-True] PASSED                            [  1%]
scipy\stats\tests\test_discrete_basic.py::test_discrete_basic[boltzmann-arg3-True] PASSED                        [  1%]
scipy\stats\tests\test_discrete_basic.py::test_discrete_basic[dlaplace-arg4-True] PASSED                         [  1%]
scipy\stats\tests\test_discrete_basic.py::test_discrete_basic[geom-arg5-True] PASSED                             [  2%]
scipy\stats\tests\test_discrete_basic.py::test_discrete_basic[hypergeom-arg6-True] PASSED                        [  2%]
scipy\stats\tests\test_discrete_basic.py::test_discrete_basic[hypergeom-arg7-False] PASSED                       [  2%]
scipy\stats\tests\test_discrete_basic.py::test_discrete_basic[hypergeom-arg8-False] PASSED                       [  3%]
scipy\stats\tests\test_discrete_basic.py::test_discrete_basic[fnch-arg9-True] Windows fatal exception: access violation

Current thread 0x000054f8 (most recent call first):
  File "C:\Users\mhaberla\Desktop\scipy\build\testenv\Lib\site-packages\scipy\stats\_discrete_distns.py", line 1300 in _rvs1
  File "C:\Users\mhaberla\Desktop\scipy\build\testenv\Lib\site-packages\scipy\stats\_discrete_distns.py", line 1311 in _rvs
  File "C:\Users\mhaberla\Desktop\scipy\build\testenv\Lib\site-packages\scipy\stats\_distn_infrastructure.py", line 1060 in rvs
  File "C:\Users\mhaberla\Desktop\scipy\build\testenv\Lib\site-packages\scipy\stats\_distn_infrastructure.py", line 3078 in rvs
  File "C:\Users\mhaberla\Desktop\scipy\build\testenv\Lib\site-packages\scipy\stats\tests\test_discrete_basic.py", line 33 in test_discrete_basic
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\python.py", line 184 in pytest_pyfunc_call
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\python.py", line 1627 in runtest
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\runner.py", line 163 in pytest_runtest_call
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\runner.py", line 256 in <lambda>
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\runner.py", line 310 in from_call
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\runner.py", line 256 in call_runtest_hook
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\runner.py", line 216 in call_and_report
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\runner.py", line 127 in runtestprotocol
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\runner.py", line 110 in pytest_runtest_protocol
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\main.py", line 338 in pytest_runtestloop
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\main.py", line 313 in _main
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\main.py", line 257 in wrap_session
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\main.py", line 306 in pytest_cmdline_main
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "C:\Users\mhaberla\Anaconda3\envs\scipydev\lib\site-packages\_pytest\config\__init__.py", line 165 in main
  File "C:\Users\mhaberla\Desktop\scipy\build\testenv\Lib\site-packages\scipy\_lib\_testutils.py", line 67 in __call__
  File "runtests.py", line 322 in main
  File "runtests.py", line 565 in <module>
```
and it doesn't go away when I add `#include <cstddef>` to `stocR.h`.

**Comment by mckib2** [comment]
Is that using the reverted branch?  What version of numpy?

**Comment by mdhaber** [comment]
No, the branch with the revert is fine. 
Before the revert I get that error during the test, which is what CI was getting before the revert. NumPy 1.19 on this machine.

**Comment by mckib2** [comment]
Okay, so mckib2:biasedurn-crash-fix has a crash during the tests.  `access violation` is a new one.  I'll have to pull my windows machine out.  Do you know if you get the same error on your mac?

File: scipy/stats/biasedurn.pyx.templ [hunk 1]
```diff
@@ -54,42 +58,68 @@ cdef class _PyWalleniusNCHypergeometric:
...
     cdef unique_ptr[StochasticLib3] c_sl3
+    IF not NPY_OLD:
+        cdef bitgen_t* bit_generator
 
     def __cinit__(self):
         self.c_sl3 = unique_ptr[StochasticLib3](new StochasticLib3(0))
+        IF NPY_OLD:
+            self.c_sl3.get().next_double = &next_double
+        ELSE:
+            self.c_sl3.get().next_double = self.bit_generator.next_double
```
Comment (mckib2) [inline scipy/stats/biasedurn.pyx.templ:97]: Might be complaining about this -- `bit_generator` is not initialized here.  I assumed that the compiler would be smart enough not to try to access the location, just point to it.  To test the hypothesis you'd move this line down to line 119 (right after the `bit_generator` is assigned).  I would fix the bad function call in the next review before trying this.

File: scipy/stats/biasedurn.pyx.templ [hunk 2]
```diff
@@ -108,9 +138,9 @@ cdef class _PyStochasticLib3:
         return rvs
 
     def FishersNCHyp(self, int n, int m, int N, double odds):
-        self.c_sl3.get().SetBitGen(make_rng())  # get default rng
+        self.HandleRng(None)  # get default rng
         return self.c_sl3.get().FishersNCHyp(n, m, N, odds)
 
     def WalleniusNCHyp(self, int n, int m, int N, double odds):
-        self.c_sl3.get().SetBitGen(make_rng())  # get default rng
+        self.InitRng(None)  # get default rng
```
Comment (mckib2) [inline scipy/stats/biasedurn.pyx.templ:145]: This is also wrong, this should be `self.HandleRng(None)`

**Comment by mdhaber** [comment]
> Okay, so mckib2:biasedurn-crash-fix has a crash during the tests. access violation is a new one. I'll have to pull my windows machine out. Do you know if you get the same error on your mac?

On macOS, commit 45b851315 builds and fails tests with the following output whether I add `#include <cstddef>` to `stocR.h` or not.
```
(scipydev) Matts-MacBook-Pro:scipy matthaberland$ python runtests.py -v -t scipy/stats/tests/test_discrete_basic.py
Building, see build.log...
Build OK (0:00:15.567269 elapsed)
============================= test session starts ==============================
platform darwin -- Python 3.7.7, pytest-5.4.1, py-1.9.0, pluggy-0.13.1 -- /opt/anaconda3/envs/scipydev/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/matthaberland/Desktop/scipy/build/testenv/lib/python3.7/site-packages/.hypothesis/examples')
rootdir: /Users/matthaberland/Desktop/scipy, inifile: pytest.ini
plugins: hypothesis-5.41.2
collected 282 items                                                            

scipy/stats/tests/test_discrete_basic.py::test_discrete_basic[bernoulli-arg0-True] PASSED [  0%]
scipy/stats/tests/test_discrete_basic.py::test_discrete_basic[betabinom-arg1-True] PASSED [  0%]
scipy/stats/tests/test_discrete_basic.py::test_discrete_basic[binom-arg2-True] PASSED [  1%]
scipy/stats/tests/test_discrete_basic.py::test_discrete_basic[boltzmann-arg3-True] PASSED [  1%]
scipy/stats/tests/test_discrete_basic.py::test_discrete_basic[dlaplace-arg4-True] PASSED [  1%]
scipy/stats/tests/test_discrete_basic.py::test_discrete_basic[geom-arg5-True] PASSED [  2%]
scipy/stats/tests/test_discrete_basic.py::test_discrete_basic[hypergeom-arg6-True] PASSED [  2%]
scipy/stats/tests/test_discrete_basic.py::test_discrete_basic[hypergeom-arg7-False] PASSED [  2%]
scipy/stats/tests/test_discrete_basic.py::test_discrete_basic[hypergeom-arg8-False] PASSED [  3%]
scipy/stats/tests/test_discrete_basic.py::test_discrete_basic[fnch-arg9-True] Fatal Python error: Segmentation fault

Current thread 0x00007fffa850f380 (most recent call first):
  File "/Users/matthaberland/Desktop/scipy/build/testenv/lib/python3.7/site-packages/scipy/stats/_discrete_distns.py", line 1300 in _rvs1
  File "/Users/matthaberland/Desktop/scipy/build/testenv/lib/python3.7/site-packages/scipy/stats/_discrete_distns.py", line 1311 in _rvs
  File "/Users/matthaberland/Desktop/scipy/build/testenv/lib/python3.7/site-packages/scipy/stats/_distn_infrastructure.py", line 1060 in rvs
  File "/Users/matthaberland/Desktop/scipy/build/testenv/lib/python3.7/site-packages/scipy/stats/_distn_infrastructure.py", line 3078 in rvs
  File "/Users/matthaberland/Desktop/scipy/build/testenv/lib/python3.7/site-packages/scipy/stats/tests/test_discrete_basic.py", line 33 in test_discrete_basic
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/python.py", line 184 in pytest_pyfunc_call
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/python.py", line 1479 in runtest
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/runner.py", line 135 in pytest_runtest_call
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/runner.py", line 217 in <lambda>
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/runner.py", line 244 in from_call
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/runner.py", line 217 in call_runtest_hook
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/runner.py", line 186 in call_and_report
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/runner.py", line 100 in runtestprotocol
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/runner.py", line 85 in pytest_runtest_protocol
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/main.py", line 272 in pytest_runtestloop
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/main.py", line 247 in _main
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/main.py", line 191 in wrap_session
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/main.py", line 240 in pytest_cmdline_main
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/opt/anaconda3/envs/scipydev/lib/python3.7/site-packages/_pytest/config/__init__.py", line 125 in main
  File "/Users/matthaberland/Desktop/scipy/build/testenv/lib/python3.7/site-packages/scipy/_lib/_testutils.py", line 67 in __call__
  File "runtests.py", line 322 in main
  File "runtests.py", line 565 in <module>
Segmentation fault: 11
(scipydev) Matts-MacBook-Pro:scipy matthaberland$ 

```

**Comment by mckib2** [comment]
@mdhaber I may have made life a little more difficult, but I have fixes in https://github.com/mdhaber/scipy/pull/53 that I'm hoping resolve these issues

**Pull request merged**